### PR TITLE
Implemented aten::rot90 operation in pytorch frontend

### PR DESCRIPTION
**Pull request by Vishwa2684**
### Details:
 - Add aten::rot90 operation
 - Added support for dynamic shapes
 - Added support for clockwise rotations (negative k)

### Tickets:
 - #24088

### CODE REVIEW

File: src/frontends/pytorch/src/op/rot90.cpp [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+#include "openvino/op/add.hpp"
+#include "openvino/op/squeeze.hpp"
+#include "openvino/op/gather.hpp"
+#include "openvino/op/divide.hpp"
+#include "openvino/op/mod.hpp"
+#include "openvino/op/add.hpp"
+#include "openvino/op/broadcast.hpp"
+#include "utils.hpp"
+
+// Can't get rank of dynamic dimension
```
Comment (mvafin) [inline src/frontends/pytorch/src/op/rot90.cpp]:
```suggestion
```
This comment make no sense there

File: src/frontends/pytorch/src/op/rot90.cpp [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+namespace ov{
+namespace frontend{
+namespace pytorch{
+namespace op{
+using namespace ov::op;
+// handle negative axes
```
Comment (mvafin) [inline src/frontends/pytorch/src/op/rot90.cpp]:
```suggestion
```
This comment make no sense there

File: src/frontends/pytorch/src/op/rot90.cpp [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+    auto tensor = context.get_input(0);
+    auto k = context.input_is_none(1) ? 1 : context.const_input<int64_t>(1);
+    const auto dims = context.input_is_none(2) ? context.mark_node(v0::Constant::create(element::i64, {2}, 
+            {0, 1})) : context.get_input(2);
+    k = (k%4+4)%4;
+    auto zero = v0::Constant::create(element::i64, Shape{}, {0});
+    auto one = v0::Constant::create(element::i64, Shape{}, {1});
+    // normalizing dims to handle negative values
+    auto shape_of = context.mark_node(std::make_shared<v3::ShapeOf>(tensor, element::i64));
+    auto rank_node = context.mark_node(std::make_shared<v3::ShapeOf>(shape_of, element::i64));
```
Comment (mvafin) [inline src/frontends/pytorch/src/op/rot90.cpp]: Please use `get_shape_rank` helper function

File: src/frontends/pytorch/src/op/rot90.cpp [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+    ));
+    // create a tensor of dims shape to hold rank
+    auto rank_bcast = context.mark_node(std::make_shared<v3::Broadcast>(
+        rank_scalar,
+        std::make_shared<v3::ShapeOf>(dims,element::i64)
+    ));
+    // normalization of dims is done here
+    auto dim_mod = context.mark_node(std::make_shared<v1::Mod>(dims,rank_bcast));
+    auto dim_mod_plus_rank = context.mark_node(std::make_shared<v1::Add>(dim_mod,rank_bcast));
+    auto normalized_dims = context.mark_node(std::make_shared<v1::Mod>(dim_mod_plus_rank,rank_bcast));
```
Comment (mvafin) [inline src/frontends/pytorch/src/op/rot90.cpp]: Please use `normalize_axis` helper function

File: src/frontends/pytorch/src/op/rot90.cpp (hunk 3 above)
Comment (Vishwa2684) [reply src/frontends/pytorch/src/op/rot90.cpp]: @mvafin  sir, I've utilized this method in my recent commit.  https://github.com/openvinotoolkit/openvino/pull/33177/commits/eb5a4ef84c7222ee89a7d2537118eb7e7a19cdc6

File: src/frontends/pytorch/src/op/rot90.cpp (hunk 4 above)
Comment (Vishwa2684) [reply src/frontends/pytorch/src/op/rot90.cpp]: @mvafin sir, I've utilized this method in my recent commit. https://github.com/openvinotoolkit/openvino/commit/eb5a4ef84c7222ee89a7d2537118eb7e7a19cdc6

**Review by mvafin — COMMENTED** [review]
Looks good to me! Ready for approval once tests pass

**Comment by Vishwa2684** [comment]
@mvafin Sir, could you please run the CI/Jenkins workflow?
Also, please let me know if any further changes are required from my side.

**Comment by Vishwa2684** [comment]
@mvafin can i use git rebase to make my branch clean, It is my first time working on a PR and I'd be grateful if you could guide me on this?

File: docs/requirements.txt [hunk 5]
```diff
@@ -23,8 +23,8 @@ packaging==25.0
 pluggy==1.6.0
 pydata-sphinx-theme==0.14.4
 Pygments==2.19.2
-pyparsing==3.3.1
-pytest==9.0.2
+pyparsing==3.2.5
```
Comment (praasz) [inline docs/requirements.txt]: Should be reverted?

File: .github/workflows/coverage.yml [hunk 6]
```diff
@@ -67,6 +67,10 @@ jobs:
             -DENABLE_OV_ONNX_FRONTEND=ON
             -DENABLE_STRICT_DEPENDENCIES=OFF
             -DENABLE_COVERAGE=ON
+            -DCMAKE_C_FLAGS="-fprofile-update=atomic"
```
Comment (praasz) [inline .github/workflows/coverage.yml]: Why is required to change this workflow?

File: src/frontends/pytorch/src/op/rot90.cpp [hunk 7]
```diff
@@ -0,0 +1,69 @@
+// Copyright (C) 2018-2025 Intel Corporation
```
Comment (praasz) [inline src/frontends/pytorch/src/op/rot90.cpp]:
```suggestion
// Copyright (C) 2018-2026 Intel Corporation
```
Use for all added files

File: docs/requirements.txt (hunk 5 above)
Comment (Vishwa2684) [reply docs/requirements.txt]: yes sir, it was an accident actually i tried to squash commits to 'update changes merge all the copilot changes in my test_rot90.py file'

File: .github/workflows/coverage.yml (hunk 6 above)
Comment (Vishwa2684) [reply .github/workflows/coverage.yml]: Sorry sir, it happened by acident while i tried to resolve merge conflicts, I'll fix that

File: src/frontends/pytorch/src/op/rot90.cpp (hunk 7 above)
Comment (Vishwa2684) [reply src/frontends/pytorch/src/op/rot90.cpp]: Done sir, 0c28eec8aaa51b55bfffde85a7e4071fcd180379

**Comment by praasz** [comment]
build_jenkins

File: src/frontends/pytorch/src/op/rot90.cpp [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+    if (k == 3) {
+        auto flipped = context.mark_node(std::make_shared<v1::Reverse>(tensor, dim0_axis, v1::Reverse::Mode::INDEX));
+        auto perm = context.mark_node(std::make_shared<v4::Range>(zero, rank_scalar, one, element::i64));
+        auto dims_perm =
+            context.mark_node(std::make_shared<v3::ScatterUpdate>(perm, normalized_dims, reverse_order, zero));
+        auto transposed = context.mark_node(std::make_shared<v1::Transpose>(flipped, dims_perm));
+        return {transposed};
+    }
+    return {tensor};
+};
```
[the remark this answers is not in the document]
Comment (Vishwa2684) [reply src/frontends/pytorch/src/op/rot90.cpp]: @mlukasze check this commit e87033c77264dcf7aa8381de0d0b1e637701cabe

File: tests/layer_tests/pytorch_tests/test_rot90.py [hunk 9]
```diff
@@ -0,0 +1,35 @@
...
+        class aten_rot90(torch.nn.Module):
+            def __init__(self, k, dims=(0, 1)):
+                super(aten_rot90, self).__init__()
+                self.k = k
+                self.dims = dims
+
+            def forward(self, x):
+                return torch.rot90(x, self.k, self.dims)
+
+        return aten_rot90(k, dims), "aten::rot90"
```
[the remark this answers is not in the document]
Comment (Vishwa2684) [reply tests/layer_tests/pytorch_tests/test_rot90.py]: @mlukasze  check this commit [e370750](https://github.com/openvinotoolkit/openvino/pull/33177/commits/e370750d7ccd4137e7ff56e4117dc6fe55644f40)

File: tests/layer_tests/pytorch_tests/test_rot90.py [hunk 10]
```diff
@@ -0,0 +1,35 @@
...
+        class AtenRot90(torch.nn.Module):
+            def __init__(self, k, dims=(0, 1)):
+                super(AtenRot90, self).__init__()
+                self.k = k
+                self.dims = dims
+
+            def forward(self, x):
+                return torch.rot90(x, self.k, self.dims)
+
+        return AtenRot90(k, dims), "aten::rot90"
```
[the remark this answers is not in the document]
Comment (Vishwa2684) [reply tests/layer_tests/pytorch_tests/test_rot90.py:33]: 0dce8e7dd19b088083a38c471d31a296f38026cc added extra cases in this commit

**Comment by Vishwa2684** [comment]
@mlukasze I've resolved the changes requested by copilot, Could you run checks?

**Comment by Vishwa2684** [comment]
@mvafin I've made the changes in rot90.cpp, test_rot90.py and I've added this operator in Pytorch FX in op_table.cpp, could you review these changes?

**Comment by p-durandin** [comment]
build_jenkins

**Comment by Vishwa2684** [comment]
@p-durandin , I developed this translator in a device which doesn't have Intel GPU. 
Could you provide me the logs for '[ie_tests_cldnn_unit_dg2_ubuntu22_release](https://openvino-ci.toolbox.iotg.sclab.intel.com/job/private-ci/job/ie/job/ie-tests-linux-ubuntu22-cldnn_unit_dg2/16986/)' workflow so that I can debug this ?

File: src/frontends/pytorch/src/op/rot90.cpp [hunk 11]
```diff
@@ -0,0 +1,76 @@
...
+using namespace ov::op;
+OutputVector translate_rot90(const NodeContext& context) {
+    num_inputs_check(context, 1, 3);
+    auto tensor = context.get_input(0);
+    const auto input_size = context.get_input_size();
+    int64_t k = (input_size < 2 || context.input_is_none(1)) ? 1 : context.const_input<int64_t>(1);
+    auto dims = (input_size < 3 || context.input_is_none(2))
+                    ? context.mark_node(v0::Constant::create(element::i64, {2}, {0, 1}))
+                    : get_input_concat_if_list(context, 2);
+    k = (k % 4 + 4) % 4;
```
Comment (praasz) [inline src/frontends/pytorch/src/op/rot90.cpp:35]: Could be explained why `+4` is required?
the result will be not same when only aply k % 4 ?

Comment (Vishwa2684) [reply src/frontends/pytorch/src/op/rot90.cpp:35]: The inner modulus returns a value between 0 to -3. And if we do direct modulus with negative number it gives a negative value, so I added +4 to handle negative values and then brought the overall value into the range of 0 to 3 using the outer modulus.

**Comment by mvafin** [comment]
build_jenkins

File: src/frontends/pytorch/src/op/rot90.cpp [hunk 12]
```diff
@@ -0,0 +1,76 @@
...
+    const auto rank = std::get<1>(shape_and_rank);
+    // normalizing dims to handle negative values
+    dims = context.mark_node(std::make_shared<v1::ConvertLike>(dims, rank));
+    const auto normalized_dims = normalize_axis(context, dims, rank);
+    // using these scalars for creating axes for flipping
+    auto dim1_index = v0::Constant::create(element::i64, {1}, {1});
+    auto dim1_axis = context.mark_node(std::make_shared<v8::Gather>(normalized_dims, dim1_index, zero));
+    auto dim0_index = v0::Constant::create(element::i64, {1}, {0});
+    auto dim0_axis = context.mark_node(std::make_shared<v8::Gather>(normalized_dims, dim0_index, zero));
+    // reverse order matrix for transpose permutation
```
[the remark this answers is not in the document]
Comment (mvafin) [reply src/frontends/pytorch/src/op/rot90.cpp:49]: I don't think this is required, but nice to have

**Comment by p-durandin** [comment]
build_jenkins

**Comment by Vishwa2684** [comment]
@p-durandin  Does any part of my code affect permissions during the execution of the follwoing workflow 
https://github.com/openvinotoolkit/openvino/actions/runs/23589802488/job/68856889736?pr=33177 ?

**Comment by p-durandin** [comment]
build_jenkins
